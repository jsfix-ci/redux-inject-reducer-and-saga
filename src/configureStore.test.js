import React from "react";
import {render} from "enzyme";
import {compose, bindActionCreators} from "redux";
import {connect} from "react-redux";
import {put, takeLatest} from "redux-saga/effects";
import {configureStore, injectReducer, injectSaga} from "./";

const getWrappedSagaComponent = prefix => {
  // constants
  const NAME_SEND = `${prefix}/NAME_SEND`;
  const NAME_SET = `${prefix}/NAME_SET`;

  // actions
  const nameSendAction = name => ({
    type: NAME_SEND,
    payload: name,
  });

  const nameSetAction = name => ({
    type: NAME_SET,
    payload: name,
  });

  // reducers
  const reducer = (state = "no-name", {type, payload}) => {
    switch (type) {
      case NAME_SET:
        return payload;
      default:
        return state;
    }
  };

  // sagas
  function* nameSetSaga(action) {
    const name = action.payload;

    yield put(nameSetAction(name));
  }

  function* saga() {
    yield takeLatest(NAME_SEND, nameSetSaga);
  }

  const withReducer = injectReducer({key: `${prefix}/testReducer`, reducer});
  const withSaga = injectSaga({key: `${prefix}/testSaga`, saga});
  const withConnect = connect(
    state => ({name: state[`${prefix}/testReducer`]}),
    dispatch => bindActionCreators({nameSend: nameSendAction}, dispatch),
  );

  // eslint-disable-next-line no-shadow
  return compose(withReducer, withSaga, withConnect)(({nameSend}) => {
    nameSend(`Name ${prefix}`);

    return null;
  });
};

const getWrappedThunkComponent = prefix => {
  // constants
  const NAME_SET = `${prefix}/NAME_SET`;

  const nameSet = name => ({
    type: NAME_SET,
    payload: name,
  });

  // reducers
  const reducer = (state = "no-name", {type, payload}) => {
    switch (type) {
      case NAME_SET:
        return payload;
      default:
        return state;
    }
  };

  // thunks
  const nameSendThunk = name => dispatch => dispatch(nameSet(name));

  const withReducer = injectReducer({key: `${prefix}/testReducer`, reducer});
  const withConnect = connect(
    state => ({name: state[`${prefix}/testReducer`]}),
    dispatch => bindActionCreators({sendName: nameSendThunk}, dispatch),
  );

  // eslint-disable-next-line no-shadow
  return compose(withReducer, withConnect)(({sendName}) => {
    sendName(`Name ${prefix}`);

    return null;
  });
};

function staticReducer(state = null) {
  return state;
}

describe("configureStore", () => {
  let store;

  beforeAll(() => {
    store = configureStore({config: staticReducer}, {config: true});
  });

  it("default store", () => {
    expect(store.getState()).toMatchSnapshot();
  });

  it("component A, with reducer and saga", () => {
    const WrappedComponent = getWrappedSagaComponent("A");

    expect(store.getState()).toMatchSnapshot();
    const component = render(<WrappedComponent />, {context: {store}});
    expect(component).toMatchSnapshot();
    expect(store.getState()).toMatchSnapshot();
  });

  it("component B, with reducer and saga", () => {
    const WrappedComponent = getWrappedSagaComponent("B");

    expect(store.getState()).toMatchSnapshot();
    const component = render(<WrappedComponent />, {context: {store}});
    expect(component).toMatchSnapshot();
    expect(store.getState()).toMatchSnapshot();
  });

  it("component C, with reducer and using thunk", () => {
    const WrappedComponent = getWrappedThunkComponent("C");

    expect(store.getState()).toMatchSnapshot();
    const component = render(<WrappedComponent />, {context: {store}});
    expect(component).toMatchSnapshot();
    expect(store.getState()).toMatchSnapshot();
  });
});
