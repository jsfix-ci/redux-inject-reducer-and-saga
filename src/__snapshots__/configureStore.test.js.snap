// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`configureStore component A, with reducer and saga 1`] = `
Immutable.Map {
  "config": true,
  "route": Immutable.Map {
    "location": null,
  },
}
`;

exports[`configureStore component A, with reducer and saga 2`] = `null`;

exports[`configureStore component A, with reducer and saga 3`] = `
Immutable.Map {
  "config": true,
  "route": Immutable.Map {
    "location": null,
  },
  "A/testReducer": "Name A",
}
`;

exports[`configureStore component B, with reducer and saga 1`] = `
Immutable.Map {
  "config": true,
  "route": Immutable.Map {
    "location": null,
  },
  "A/testReducer": "Name A",
}
`;

exports[`configureStore component B, with reducer and saga 2`] = `null`;

exports[`configureStore component B, with reducer and saga 3`] = `
Immutable.Map {
  "config": true,
  "route": Immutable.Map {
    "location": null,
  },
  "A/testReducer": "Name A",
  "B/testReducer": "Name B",
}
`;

exports[`configureStore component C, with reducer and using thunk 1`] = `
Immutable.Map {
  "config": true,
  "route": Immutable.Map {
    "location": null,
  },
  "A/testReducer": "Name A",
  "B/testReducer": "Name B",
}
`;

exports[`configureStore component C, with reducer and using thunk 2`] = `null`;

exports[`configureStore component C, with reducer and using thunk 3`] = `
Immutable.Map {
  "config": true,
  "route": Immutable.Map {
    "location": null,
  },
  "A/testReducer": "Name A",
  "B/testReducer": "Name B",
  "C/testReducer": "Name C",
}
`;

exports[`configureStore default store 1`] = `
Immutable.Map {
  "config": true,
  "route": Immutable.Map {
    "location": null,
  },
}
`;
